presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexSelectionStrategy.java

```diff
@@ -0,0 +1,257 @@
+            return none();
+        }
+        else {
+            IndexSelectionStrategy indexSelectionStrategy = new IndexSelectionStrategy(Type.CUSTOM);
+            indexSelectionStrategy.tables = ImmutableSet.copyOf(tables);
+            return indexSelectionStrategy;
+        }
+    }
+
+    static IndexSelectionStrategy thresholdWithPercent(int maxPercent)
+    {
+        IndexSelectionStrategy indexSelectionStrategy = new IndexSelectionStrategy(Type.THRESHOLD);
+        indexSelectionStrategy.maxPercent = maxPercent;
+        return indexSelectionStrategy;
+    }
+
+    static IndexSelectionStrategy thresholdWithRows(int maxRows)
+    {
+        IndexSelectionStrategy indexSelectionStrategy = new IndexSelectionStrategy(Type.THRESHOLD);
+        indexSelectionStrategy.maxRows = maxRows;
+        return indexSelectionStrategy;
+    }
+
+    public boolean isContained(@Nonnull SchemaTableName table)
+    {
+        if (this == AUTO || this.type == Type.THRESHOLD) {
+            return true;
+        }
+        if (this == NONE) {
+            return false;
+        }
+        return tables.contains(table);
+    }
+
+    public String backToString()
+    {
+        if (type == Type.AUTO || type == Type.NONE) {
+            return type.getCode();
+        }
+        if (type == Type.THRESHOLD) {
+            String str = type.getCode();
+            str += maxPercent > 0 ? ":" + maxPercent + "%" : "";
+            str += maxRows > 0 ? ":" + maxRows : "";
+            return str;
+        }
+        Optional<String> str = tables.stream().map(SchemaTableName::toString).reduce((a, b) -> a + "," + b);
+        return "[" + str.orElse("") + "]";
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static IndexSelectionStrategy parse(Optional<String> currentConnectionSchema, String hintKey, String hintValue)
+    {
+        if (StringUtils.isBlank(hintValue)) {
+            return IndexSelectionStrategy.none();
+        }
+        String hint = hintValue.trim().toLowerCase(Locale.ENGLISH);
+        if (Type.AUTO.getCode().equals(hint)) {
+            return IndexSelectionStrategy.auto();
+        }
+        if (Type.NONE.getCode().equals(hint)) {
+            return IndexSelectionStrategy.none();
+        }
+        String pattern = "^" + Type.THRESHOLD.getCode() + ":\\s*(\\d+)\\s*(%)?$";
+        if (hint.matches(pattern)) {
+            String err = "Invalid 'threshold' hint value '" + hintValue + "' of hint key '" + hintKey + "', regular pattern='" + pattern + "'";
+            Matcher m = Pattern.compile(pattern).matcher(hint);
+            checkState(m.find(), err);
+
+            int integer = Integer.parseInt(m.group(1));
+            String percent = m.group(2);
+
+            if (percent != null) {
+                if (integer < 1 || integer > MAX_VALUE_OF_INDEX_FIRST_MAX_PERCENT) {
+                    err = "Invalid 'threshold:${maxPercent}%' hint value '" + hintValue + "' of hint key '" + hintKey + "', "
```

Comment:
shall we inline error message creation in IllegalArgumentException? instead of having an err variable?